### DISCUSSION TITLE
How to observe PopupController is currently presenting or in a hidden state

### DISCUSSION THREAD

**Discussion by dfeler** [LeoNatan/LNPopupController#503] [General] [answered]
Each time I have the following execute in `ViewController1`: 
`tabBarController?.presentPopupBar(withContentViewController: navigationController, animated: true, completion: nil)` it trigger the ViewWillAppear and ViewWillDisappear properties to execute in `ViewController2` presented using the library. This is not great for my program as I have other function in those functions I need executed only when the PopupController is actually being presented on the entire entire screen in and `openPopup` state.

Is it possible to constantly observe if a PopupController is an `openPopup` state to execute a function each time?

I appreciate the help.

**Comment by LeoNatan** [comment, marked as answer]
Please stop opening question “issues”, as stated by the template, use Discussions for these. 

There is a delegate you can assign to listen to state changes. 

https://github.com/LeoNatan/LNPopupController/blob/e8d646a44a8569d23c7418f1de39de2e58cb90da/LNPopupController/LNPopupController/UIViewController%2BLNPopupSupport.h#L109

**Comment by dfeler** [comment]
Can you please show an example of how this can be used in Swift? I tried doing:
This in ViewController2 (the presenting controller) 

```
class ViewController2: UITableViewController, UIGestureRecognizerDelegate, CLLocationManagerDelegate, LNPopupPresentationDelegate {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        popupPresentationDelegate = self
    }
    
    func popupPresentationControllerDidDismissPopupBar(_ popupPresentationController: UIViewController, animated: Bool) {
        print("Did Dismiss")
    }
    func popupPresentationControllerDidPresentPopupBar(_ popupPresentationController: UIViewController, animated: Bool) {
        print("Did Present")
    }
    
}
```
Did Dismiss / Did Present are not printing.

Is anything missing? I appreciate the help.

**Reply by LeoNatan** [reply]
You are setting it on the wrong object. The property should be set on the controller you present the popup on.

**Reply by dfeler** [reply]
Ok now I have set up the following in ViewController1 (the controller that presents the popup) the same way:

```
class ViewController2: UITableViewController, UIGestureRecognizerDelegate, CLLocationManagerDelegate, LNPopupPresentationDelegate {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        popupPresentationDelegate = self
    }
    
    func popupPresentationControllerDidDismissPopupBar(_ popupPresentationController: UIViewController, animated: Bool) {
        print("Did Dismiss")
    }
    func popupPresentationControllerDidPresentPopupBar(_ popupPresentationController: UIViewController, animated: Bool) {
        print("Did Present")
    }
    
}
```

It is unfortunately still not printing Did Dismiss / Did Present? Is popupPresentationDelegate = self set correctly in viewdidload? Am I missing anything else to get those functions to execute?

**Reply by LeoNatan** [reply]
How do you present the popup?

**Reply by dfeler** [reply]
func openpopUp() {
        let stopDetails = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(withIdentifier: "ViewController2") as! ViewController2
        let navigationController = UINavigationController(rootViewController: ViewController2)
        navigationController.popupItem.title = "ViewController2"
        navigationController.popupItem.subtitle = "ViewController2"
        
        let popupNextButtonItem = UIBarButtonItem(image: UIImage(named: "arrowcircleGlyph"), style: .plain, target: self, action: #selector(openTab))
            
        navigationController.popupItem.trailingBarButtonItems = [popupNextButtonItem]
        tabBarController?.popupContentView.popupCloseButtonStyle = .none
        tabBarController?.popupInteractionStyle = .none
        navigationController.popupBar.backgroundColor = .white
        navigationController.popupBar.inheritsAppearanceFromDockingView = false
        navigationController.popupBar.tintColor = .white
        navigationController.popupContentView.isTranslucent = false
        navigationController.popupContentView.isOpaque = true
        navigationController.popupContentView.popupInteractionGestureRecognizer.delegate = self
        
        //Open popup
        tabBarController?.presentPopupBar(withContentViewController: navigationController, animated: true, completion: nil)
        
    }


This function is executed on didSelect in a tableView Controller

**Reply by LeoNatan** [reply]
I don’t understand. So set the delegate to the tab bar controller.

**Reply by dfeler** [reply]
Thank you for the help.
